keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1194 @@
+
+	// Status represents the most recently observed status of the evacuation.
+	// Populated by the current evacuator.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Status EvacuationStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// EvacuationSpec is a specification of an Evacuation.
+type EvacuationSpec struct {
+	// PodRef references a pod that is subject to evacuation.
+	// This field is required and immutable.
+	// +optional
+	PodRef *PodReference `json:"podRef,omitempty" protobuf:"bytes,1,opt,name=podRef"`
+
+	// AcceptDeadlineSeconds is a maximum amount of time an evacuator should take to respond to an
+	// evacuation. It should respond by setting .status.accepted to true or false, and providing
+	// .status.evacuatorRef and .status.evacuationProgressTimestamp. The evacuator should then
+	// periodically update .status.evacuationProgressTimestamp.
+	// If the .status.accepted is not updated within the duration of .spec.acceptDeadlineSeconds,
+	// the evacuated pod will be evicted using the Eviction API.
+	//
+	// The minimum value is 60 (1m) and the maximum value is 3600 (1h).
+	// The value must be less than or equal to .spec.progressDeadlineSeconds.
+	// This field is required and immutable.
+	AcceptDeadlineSeconds int32 `json:"acceptDeadlineSeconds" protobuf:"varint,2,opt,name=acceptDeadlineSeconds"`
+	
+	// ProgressDeadlineSeconds is a maximum amount of time an evacuator should take to report on a
+	// progress by updating the .status.evacuationProgressTimestamp.
+	// If the .status.evacuationProgressTimestamp is not updated within the duration of
+	// .spec.progressDeadlineSeconds, the evacuated pod will be evicted using the Eviction API.
+	//
+	// The minimum value is 60 (1m) and the maximum value is 43200 (12h).
+	// The value must be greater than or equal to .spec.acceptDeadlineSeconds.
+	// This field is required and immutable.
+	ProgressDeadlineSeconds int32 `json:"progressDeadlineSeconds" protobuf:"varint,3,opt,name=progressDeadlineSeconds"`
+}
+
+// PodReference contains enough information to locate the referenced pod.
+type PodReference struct {
+	// Name of the pod.
+	// This field is required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// Namespace of the pod.
+	// This field is required.
+	Namespace string `json:"namespace" protobuf:"bytes,2,opt,name=namespace"`
+	// UID of the pod.
+	// This field is required.
+	UID string `json:"uid" protobuf:"bytes,3,opt,name=uid"`
+}
+
+// EvacuationStatus represents the most recently observed status of the evacuation.
+type EvacuationStatus struct {
+	// EvacuatorRef references the evacuator of the current evacuation. 
+	// This field should be set when Accepted field is set.
+	// +optional
+	EvacuatorRef *EvacuatorReference `json:"evacuatorRef,omitempty" protobuf:"bytes,1,opt,name=evacuatorRef"`
```

Comment:
How will this work if two different things both want to see a Pod removed
(eg: the Node is gracefully shutting down, part way through a rollout triggered by a ConfigMap update)

Reply:
This has to be resolved by the owning controller (evacuator). Also, this KEP needs to be updated to include a flow for multiple evacuators as mentioned in https://github.com/kubernetes/enhancements/pull/4213#discussion_r1543218466

Reply:
resolved in the recent update: https://github.com/kubernetes/enhancements/pull/4565#issuecomment-2045722703